### PR TITLE
Add flake.lock & workflow to update it

### PR DESCRIPTION
**Pull request by i077**
### Description

This PR implements suggestions from #1913, adding a flake.lock file so we can `nix run github:nix-community/home-manager` without any hiccups. The nixpkgs input is currently set to track `nixos-unstable`. If you want to test this out, you can run `nix run github:nix-community/home-manager/pull/1934/head`, which should give you the help text.

To keep this file up to date, the PR also adds a GitHub workflow that
- Runs every Sunday and Wednesday at 03:51 (I just picked a random time)
- Runs `nix flake update --commit-lock-file`, committing as a nonexistent user called `flakebot`
- Pushes this commit ~~directly to master~~ to a new branch called `flake-update` with a custom token `FLAKEBOT_TOKEN` (this needs to be added to the repo)
- Opens a PR (or updates an existing PR) to merge this commit to master. This is done with the same custom token so that other workflows can be run against the PR.

Also updates the README to change the section on a flake-based config a bit.

I also wanted to modify the testing workflow to actually run home-manager's tests, but was having a bit of trouble exposing the tests under the `checks` output of the flake, so I will leave that out for now. ~~This means that the update workflow will just regularly push updates to master, whether they pass tests or not.~~

cc @berbiche since we were talking about this in the other PR.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [ ] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: .github/workflows/update.yml
Comment (kclejeune) [inline .github/workflows/update.yml:25]: Might be a good idea to verify that `nix run github:nix-community/home-manager` will run with exit code 0 before proceeding with the push action

Comment (i077) [reply .github/workflows/update.yml:25]: Good point, I can add that as an extra step.

**Comment by teto** [comment]
at some point it was discussed to open a PR instead of an automatic merge. I preferred that scenario, is that doable through github actions ? nevermind the checks, they can be added later

**Comment by i077** [comment]
I can have the workflow `checkout -B` a new branch and open a new PR, that's what @kclejeune and I do on our nixos configs. My only concern was that I didn't want the repo to just have a bunch of flake update PRs, but if you're okay with that I can implement it.

**Comment by teto** [comment]
> edit: I should clarify: no more than 1 flake update PR will be open at a time, but it will require manual merging each time it gets opened (at most twice a week).

That seems the best option, would it close and open a new PR or force-push to the current one ?

**Comment by i077** [comment]
If the update action runs and there's already an existing PR, it will just force-push that branch. The `-B` switch in the checkout command basically "force creates" a new branch, even if it already existed in the remote, so the previous update commit just gets overwritten.

I'll update the workflow later today.

**Comment by kclejeune** [comment]
Can you make a note that this approach will require adding `FLAKEBOT_TOKEN` as a repository secret?

File: README.md
Comment (teto) [inline README.md:336]: this is potentially harmful, advanced users can use it but we should default to the mainstream method.

Comment (kclejeune) [reply README.md:336]: could you elaborate on how this is potentially harmful?

Comment (teto) [reply README.md:336]: home-manager is guaranteed to work with the nixpkgs in its lock file, it has not been tested with the overriden version so there is a risk of failure no ?

Comment (i077) [reply README.md:336]: IMO users typically expect their entire configuration to be evaluated against a single instance of `nixpkgs` unless they specify otherwise, so I figured overriding the input would be a sensible thing to include here.

Comment (teto) [reply README.md:336]: I disagree, that's not how flakes are conceived (it doesn't default to a single nixpkgs) and I think users prefer something that work rather than shaving off a few MB on the disk. It's still possbile but shouldn't be the default I think.

Comment (i077) [reply README.md:336]: Right, but I think that would just introduce another source of confusion for users. If they update their nixpkgs input, they're _probably_ expecting to see that update reflected across all their configs.

Additionally, we're going to be tracking nixos-unstable, but what about the people whose configs use the stable branch? It's probably not a good idea for them to mix up stable with unstable (a bit more consequential than saving a few MB...), so they'd _need_ to explicitly override the nixpkgs input.

I suppose I can add a disclaimer instead that the user's home-manager config will use a different instances of nixpkgs than their nixos/darwin config instead, and just add a note on how to override that.

Comment (teto) [reply README.md:336]: if they track the stable version then they should use the home-manager flake from the stable branch. In any case, as long as they use the locked flakes, any combination should work (I think), it's only if they replace the locked inputs that they can have issues.

Comment (teto) [reply README.md:336]: we could mention the `follow` trick in a "Flake" page over the wiki https://github.com/nix-community/home-manager/wiki

Comment (i077) [reply README.md:336]: Should we then set up periodic updates for the stable branch as well to track `nixos-20.09` (and future versions)?

Comment (teto) [reply README.md:336]: That would probably be a good idea but I think we could merge already as is. I would actually merge if I could setup the FLAKEBOT_TOKEN. What are the restrictions to set it ?

Comment (i077) [reply README.md:336]: It just needs the repo permission to have push access to its flake-update branch.

Comment (i077) [reply README.md:336]: If you have write access to the repo I think you should be able to create secrets. Just note that if you create a personal access token, the flake update PRs will be opened under your name, like in [this PR](https://github.com/i077/system/pull/38).

Comment (terlar) [reply README.md]: Note that such a token will give access to all the repos for which the user have access, as there is no way to create scoped personal access tokens. This is discussed here with a few possible solutions:
https://github.com/peter-evans/create-pull-request/issues/48

As I understand it:
- create a new user for the token and scope it down in that way
- Create a custom SSH key (which can configure access per repo level)
- Create the PR from a fork 

The reason the token is needed at all is to workaround workflows not being triggered when PR:s are opened by the default GH Actions user.

Comment (kclejeune) [reply README.md]: Another solution might be to just have an organization-level secret token for this? This would also allow the same workflow to be used in other flake projects.

Comment (berbiche) [reply README.md]: @rycee thoughts on having on this automated flake.lock dependencies update?

Comment (rycee) [reply README.md]: Sorry, I don't have much sensible input to give since I don't really know anything about flakes and even less about whether automatic lock file updates are a good idea 🙁 So I'll have to leave it up to you, let me know if there is anything I need to do with secrets and such.

If I do have to say something then I guess it would be best to have a dedicated bot account as the commit and PR author. I.e., something along the lines of @terlar's suggestion.

In the longer run, however, wouldn't the best be to have something like dependabot, which acts as a repository app? Perhaps there have already been work done in that area? Especially since I assume something like that would also be useful for the wider Nix community.

File: .github/workflows/update.yml
Comment (rycee) [inline .github/workflows/update.yml:15]: Perhaps these lines should be indented one more step?

File: README.md
Comment (kclejeune) [reply README.md]: > In the longer run, however, wouldn't the best be to have something like dependabot, which acts as a repository app? Perhaps there have already been work done in that area? Especially since I assume something like that would also be useful for the wider Nix community.

I'm working on a "flakebot" Github App to do this exact thing, but won't be able to do much on it until my semester is over in another couple weeks. For the time being, I think the workflow is still a good idea, although I still prefer an organization level flakebot token so that similar workflows could be used in other projects if necessary.

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:15]: Good catch, thanks.

File: README.md
Comment (rycee) [reply README.md]: @kclejeune Right, yeah, I had a look in the organization settings and it seem possible to create a secret that you can tie to specific repos, So that seems like a nice idea. But I guess the token should belong to the flakebot user? Who will manage that user?

@Mic92 Sorry to drag you in here but I wonder if you know of anybody else in the nix-community organization being interested in periodic CI pipeline that updates their project's `flake.lock` file? We're discussing here to create such for HM and the suggestion of @kclejeune is to add a token for a "flakebot" account as an organization level secret so that other projects in nix-community could have the possibility of adding such a pipeline.

Comment (kclejeune) [reply README.md]: @rycee I created a dummy flakebot user with @i077 a while back, and we ended up deleting it because it seems unnecessary. The token belongs to `nix-community`, not to a particular user account. Whoever owns this organization can enable it in repositories where it seems fitting.

Comment (rycee) [reply README.md]: @kclejeune, maybe I'm confused. I don't see any way to create a token for an organization. What I can see is a way to create a secret, but then I have to provide a value. Perhaps I don't have sufficient access to the nix-community organization? In the documentation I found https://docs.github.com/en/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization but it seems to be limited to `GITHUB_TOKEN`.

Comment (kclejeune) [reply README.md]: > @kclejeune, maybe I'm confused. I don't see any way to create a token for an organization. What I can see is a way to create a secret, but then I have to provide a value. Perhaps I don't have sufficient access to the nix-community organization? In the documentation I found https://docs.github.com/en/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization but it seems to be limited to `GITHUB_TOKEN`.

I think this may be restricted to org admins? I could also just be mistaken and it's just not possible at all.

I was under the impression that you could generate something analogous to a personal access token at the org level. I'll setup an organization under my own account and see if I can figure it out. Otherwise, the dummy account is probably the way to go.

Comment (i077) [reply README.md]: I don't think this is possible, personal access tokens are per-user only, not per-organization. I am an admin of another GitHub org and I see no such option to create a token at the org level.

File: .github/workflows/update.yml
Comment (asymmetric) [inline .github/workflows/update.yml:46]: Why do we need this?

File: .github/workflows/update.yml
Comment (asymmetric) [inline .github/workflows/update.yml:47]: Maybe:

```suggestion
          PULL_REQUEST_TITLE: [`flakebot`] Update flake inputs
```

File: README.md
Comment (asymmetric) [inline README.md:356]:
```suggestion
Note that `home-manager` uses its own nixpkgs input, which is updated
```

File: README.md
Comment (asymmetric) [inline README.md:357]: Do we have a link to the wiki section?

File: flake.nix
Comment (asymmetric) [inline flake.nix:34]: This is a nit, but why do we need to use a custom function when `nixpkgs.legacyPackages.${system}` would suffice?

**Comment by asymmetric** [comment]
> This PR implements suggestions from #1913, adding a flake.lock file so we can nix run github:nix-community/home-manager without any hiccups.

If this is the only reason, wouldn't it make more sense to fix it on nix's side? Why does `nix run` try to save a lock-file in the first place? And shouldn't this be configurable somewhere?

File: flake.nix
Comment (i077) [reply flake.nix:34]: Figured it was a more succinct expression that was being used elsewhere anyways. I can revert this particular change, though.

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml]: Ah, good catch. Thanks.

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:46]: This will open a new pull request against the `master` branch, or, if one exists already, update the existing one. `PASS_ON_ERROR` is set to true since this action will always be run, regardless of whether the flake was updated during the workflow. So if there were no changes, this step would fail, but we don't want the workflow to fail because of that.

File: README.md
Comment (i077) [reply README.md:356]: Oops, another good catch. Thanks!

File: README.md
Comment (i077) [reply README.md:357]: Hasn't been written yet, but I don't have access to write the page.

File: README.md
Comment (Mic92) [reply README.md]: @rycee Yes. Please open a ticket at https://github.com/nix-community/infra

File: README.md
Comment (rycee) [reply README.md:357]: It should go in the manual instead of some wiki. I believe there is some work documenting flakes in HM in #1697.

File: .github/workflows/update.yml
Comment (asymmetric) [reply .github/workflows/update.yml:28]: Did you end up adding this? Can't find it anywhere

File: .github/workflows/update.yml
Comment (asymmetric) [reply .github/workflows/update.yml:46]: Can we instead check if there was any update, and optionally skip this step? Because I'm not sure if we should have a blanked ignore of all errors. Could be hard to notice when there are actual issues.

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:28]: The PR will list the updated inputs (actually, it just inserts the auto-generated commit message), so it won't be as surprising. I think it's better that we update all inputs if more get added so we don't have a situation where inputs haven't been updated in weeks or months.

File: README.md
Comment (i077) [reply README.md:357]: Alright, so I should just take out that sentence entirely? Or leave a reference to a to-be-written section of the manual?

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:46]: Hmm, this will be a lot harder than just checking `commit_info`, since that is literally the commit info of HEAD, regardless of whether a flake update was committed or not. I suppose I can check if a commit was made by comparing against [`github.sha`](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context), though.

File: README.md
Comment (asymmetric) [reply README.md:357]: I would just remove the sentence. When we'll have docs (soon<sup>TM</sup>), we'll update the README.

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:47]: Maybe "[`automation`]" for a more general term?

Comment (asymmetric) [reply .github/workflows/update.yml:47]: Sounds good!

File: README.md
Comment (teto) [reply README.md]: can we fix this ? that's the last issue before merging I think ? @Mic92 you seem to have the necessary rights, haven't you ?

**Comment by teto** [comment]
what is blocking ? looking forward to this to make lockstep updates in home-manager. Right now, when the nixos-unstable channel updates it can very well break home-manager. It would be better if we could choose what version of nixos-unstable works well via the flake.

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:46]: After thinking about it some more, I think a better way to do this would be to ask the author of the action to let us specify to pass just if there are no new commits, that way I don't have to "reinvent the wheel". I can open an issue for that.

**Comment by Mic92** [comment]
> what is blocking ? looking forward to this to make lockstep updates in home-manager. Right now, when the nixos-unstable channel updates it can very well break home-manager. It would be better if we could choose what version of nixos-unstable works well via the flake.

From my understanding it requires a github token still? I can add one on every level.

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml]: Action author has suggested that I just implement this logic in this workflow's step, so I have pushed 0be2c37fd5beeaf60366679316f350ab4f5fef61. @asymmetric can you check if this looks right?

File: .github/workflows/update.yml
Comment (asymmetric) [inline .github/workflows/update.yml:31]:
```suggestion
          cat >> $GITHUB_ENV <<EOS
          commit_info<<EOF
          git log --format=%B -1
          EOS
```

Or alternatively, inlining with `echo -e 'commit_info<<EOF\ngit.. > $GITHUB_ENV`.

I think both of these are cleaner.

File: .github/workflows/update.yml
Comment (asymmetric) [inline .github/workflows/update.yml:35]:
```suggestion
          echo "commit_id=$(git rev-parse HEAD)" >> $GITHUB_ENV
```

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:31]: Does this work for multiline output? If so then it should be fine.

File: .github/workflows/update.yml
Comment (asymmetric) [inline .github/workflows/update.yml:45]:
```suggestion
        # only create a pull request if there are new commits
        if: ${{ env.commit_id != github.sha }}
```

File: .github/workflows/update.yml
Comment (i077) [reply .github/workflows/update.yml:31]: Alright, seems like it works fine. I'll commit these changes.

**Comment by Mic92** [comment]
If you need any organisational changes please drop a message in #nix-community:bethselamin.de or
#tmp-nix-community:numtide.com

File: .github/workflows/update.yml
Comment (asymmetric) [reply .github/workflows/update.yml]: I don't think we were looking to do multiline output anyway, IIUC. We just wanted to assign a value to a variable, correct? As explained [here](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable).

If not, then I might have misunderstood the purpose of your code.

Comment (i077) [reply .github/workflows/update.yml]: `git log --format=%B -1` will print the last commit's message, which will span multiple lines in this case. This is so that in the next step we can insert the commit message in the PR body, like you see [here](https://github.com/i077/system/pull/48).

Comment (asymmetric) [reply .github/workflows/update.yml]: Ah OK, I get it now. Well, the syntax we switched to doesn't preserve newlines AFAICT. :/

Comment (i077) [reply .github/workflows/update.yml]: I tried it in the terminal and it works on my end: 
image

Comment (asymmetric) [reply .github/workflows/update.yml]: Doesn't work on my system :confused: 

20210628_18h59m23s_grim

Comment (asymmetric) [reply .github/workflows/update.yml]:
```
[asymmetric@tachikoma:~/code/wasmer]$ bash --version
GNU bash, version 4.4.23(1)-release (x86_64-unknown-linux-gnu)
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

Comment (i077) [reply .github/workflows/update.yml]: Ah, I guess it works on zsh then. Shell scripting 🙄

**Comment by i077** [comment]
Has a token been added as a secret to this repo? What else needs to be done to move this forward?

**Comment by Mic92** [comment]
> Has a token been added as a secret to this repo? What else needs to be done to move this PR forward?

No. I have received no token so far. I think the best solution would be if one of the home-manager maintainer create a bot account that we add to the nix-community organisation.

**Comment by mainrs** [comment]
Wouldn't the correct behavior be to track the latest stable branch instead of `nixos-unstable`? And if a new release it cut, the channel is updated again :thinking:

**Comment by kclejeune** [comment]
> Wouldn't the correct behavior be to track the latest stable branch release (candidate) instead of `nixos-unstable`? And if a new release it cut, the channel is updated :thinking: 

no, home manager's master branch tracks nixpkgs-unstable. There's a separate branch that tracks nixos-[stable]

**Comment by mainrs** [comment]
> > Wouldn't the correct behavior be to track the latest stable branch release (candidate) instead of `nixos-unstable`? And if a new release it cut, the channel is updated thinking
> 
> no, home manager's master branch tracks nixpkgs-unstable. There's a separate branch that tracks nixos-[stable]

Ah, thanks! I wasn't aware of that!

**Comment by teto** [comment]
@rycee could we move forward with this ?

**Comment by berbiche** [comment]
We need to decide what strategy to use for release branches and how many release branches will receive these updates.

We could drop the lock file in release branches but we would be required to use `--no-write-lock-file` when using HM.
If we go for this solution, we can merge this PR as soon as the missing token is added.

If we ever needed to revert the changes or delete the lock file, the user's flake would automatically replace the `nixpkgs` used by Home Manager in its lockfile.

**Comment by Atry** [comment]
When I tried to execute:
```
nix run github:nix-community/home-manager --no-write-lock-file
```
I got:
```
warning: not writing modified lock file of flake 'github:nix-community/home-manager':
• Added input 'nixpkgs':
    'github:NixOS/nixpkgs/b283b64580d1872333a99af2b4cef91bb84580cf' (2022-05-01)
error: attribute 'defaultApp.x86_64-linux' should have type 'derivation'
```

However, if I execute:
```
nix shell github:nix-community/home-manager --no-write-lock-file -c home-manager
```
It works well.

I think we might also want to add a `defaultApp` to `flake.nix`

**Comment by terlar** [comment]
This is already provided (`defaultApp`). What you are seeing is a recent change in Nix 2.8 where the defaultApp needs to be a derivation as opposed to an attribute set that was the convention before.

**Comment by i077** [comment]
@DeterminateSystems have since released a [GitHub Action](https://github.com/DeterminateSystems/update-flake-lock) that does what we would want, and it's probably better for home-manager to use that rather than maintain its own workflow. I can update this workflow to using that action instead.

It will use a default automated token for opening update PRs, but we'll still need to provide a personal auth token if we want to run tests against these PRs.

**Comment by ncfavier** [comment]
https://github.com/nix-community/home-manager/pull/2971 adds a `flake.lock`, so I guess it's blocked on this.

> @DeterminateSystems have since released a [GitHub Action](https://github.com/DeterminateSystems/update-flake-lock) that does what we would want, and it's probably better for home-manager to use that rather than maintain its own workflow. I can update this workflow to using that action instead.

Yes, I think that's better.

> It will use a default automated token for opening update PRs, but we'll still need to provide a personal auth token if we want to run tests against these PRs.

This is currently not needed as the CI tests don't use the flake at all (maybe they should...).

> This is already provided (`defaultApp`). What you are seeing is a recent change in Nix 2.8 where the defaultApp needs to be a derivation as opposed to an attribute set that was the convention before.

Fixed in https://github.com/nix-community/home-manager/pull/2971, see https://github.com/nix-community/home-manager/pull/2442#issuecomment-1133670487.

--------

About the interval, I think twice a week is excessive considering that the updates require human intervention. Once every one or two weeks seems about right, maybe less often for release branches if we choose to update them (maybe release branches should follow the appropriate `nixos-` release branch instead of `unstable`?).

**Comment by i077** [comment]
> This is currently not needed as the CI tests don't use the flake at all (maybe they should...).

My hope is that we'd later port rycee's home-manager testing framework to support flakes in a later PR. I was working on that very briefly, but stopped a while ago because of schoolwork.

> About the interval, I think twice a week is excessive considering that the updates require human intervention. Once every one or two weeks seems about right, maybe less often for release branches if we choose to update them (maybe release branches should follow the appropriate nixos- release branch instead of unstable?).

I agree that release branches should follow the corresponding branch in nixpkgs. In particular, I think 
- `master` should follow `nixpkgs-unstable` since home-manager is developed against that branch anyway, and
- `release-xx.yy` should follow `nixos-xx.yy`. 

I suppose that means, going forward from `release-21.11`, we should change the nixpkgs input in `flake.nix` to follow the appropriate branch and cherry-pick backports on top of that as usual. (Cutting new release branches would involve updating flake.nix at that point as well.)

However, that would mean maintainers would have to review 2 PRs per run of this workflow (for master and current release branch).

**Comment by Mic92** [comment]
I added a secret for a github bot to this account. If you use the `GH_TOKEN_FOR_UPDATES` secret than you will get a token of this bot https://github.com/home-manager-bot
This bot can create branches on this repo

**Comment by Mic92** [comment]
The access of this bot is managed with terraform here: https://github.com/Mic92/dotfiles/blob/master/terraform/modules/github-push-bot/main.tf and here https://github.com/Mic92/dotfiles/blob/master/terraform/github-bots/main.tf

**Comment by ncfavier** [comment]
https://github.com/nix-community/home-manager/pull/2971 was merged, we have a flake.lock now.

**Comment by i077** [comment]
Thank you @ncfavier and @Mic92 for your help. A lot has happened since I opened this PR, so later today I will rebase my branch onto the tip of the master branch and redo most of the work so that it only adds the update workflow, since we now have a flake.lock.

As I mentioned somewhere earlier in this PR, future work will involve porting home-manager's tests to the `checks` output of flake.nix, so we can run CI tests against flake updates.

**Comment by i077** [comment]
Alright, branch rebased. Since most of the work in the rest of the old commits was done elsewhere, this PR now just implements that update workflow. PRs will be tagged with the "dependencies" label (like dependabot does) to distinguish it from other PRs.

File: .github/workflows/update-flake.yml
Comment (ncfavier) [inline .github/workflows/update-flake.yml:5]: I still think twice a week is excessive and will lead to PRs piling up. How about once a week?

File: .github/workflows/update-flake.yml
Comment (ncfavier) [inline .github/workflows/update-flake.yml:18]: We probably don't need this.

File: .github/workflows/update-flake.yml
Comment (kclejeune) [reply .github/workflows/update-flake.yml:5]: It wouldn't really pile up, right? It would just update an already open PR. 

It seems more important to have frequent updates available, even if the PR is sometimes left open for a while.

Comment (ncfavier) [reply .github/workflows/update-flake.yml:5]: Ah, it seems you're right. In that case it's probably fine.

File: .github/workflows/update-flake.yml
Comment (i077) [reply .github/workflows/update-flake.yml:18]: You're probably right, since this is just an update job. It wouldn't hurt to have though, right? I noticed nix-community/infra has this [in their update workflow](https://github.com/nix-community/infra/blob/fa549f5e591a41b317978023bad7b2ea91503e19/.github/workflows/flake-updates.yml#L17).

Comment (ncfavier) [reply .github/workflows/update-flake.yml:18]: I'd rather avoid giving unnecessary privileges to things, and I suspect people are cargo culting this from the update-flake-lock [README](https://github.com/DeterminateSystems/update-flake-lock#example).

Comment (i077) [reply .github/workflows/update-flake.yml:18]: That's fair. I have nothing against removing it, so I'll just take it out.

File: .github/workflows/update-flake.yml
Comment (ncfavier) [inline .github/workflows/update-flake.yml:18]: FWIW the action does this by default

Comment (i077) [reply .github/workflows/update-flake.yml:18]: Ah, didn't know they added that. Thanks.

**Review by ncfavier — APPROVED** [review]

**Comment by ShamrockLee** [comment]
The tests would be available through the `legacyPackages` output after #3024 is merged.
It can be run automatically by
```console
$ nix develop .#tests.run.all
```

**Review by berbiche — APPROVED** [review]

File: README.md
Comment (berbiche) [reply README.md]: I didn't even notice I had a stale comment in the UI

**Comment by teto** [comment]
tempted to merge this (and if it breaks stuff we'll deal with it). Any objection ?

**Comment by i077** [comment]
I'm fine with that. I will just draw attention to https://github.com/nix-community/home-manager/pull/1934#issuecomment-1134702268, in particular whether release branches should receive a similar update workflow, perhaps in a later PR:

> I suppose that means, going forward from release-21.11, we should change the nixpkgs input in flake.nix to follow the appropriate branch and cherry-pick backports on top of that as usual. (Cutting new release branches would involve updating flake.nix at that point as well.)
>
> However, that would mean maintainers would have to review 2 PRs per run of this workflow (for master and current release branch).

We don't have to address this now, but perhaps if we merge this PR we should consider doing the above before the next NixOS release.

**Pull request merged**
